File: src/model_story/parsers/fastai.py
from __future__ import annotations

import contextlib
import re

from model_story.models import RawMetric
from model_story.parsers.base import ParserContext
from model_story.parsers.registry import register_parser

# FastAI tabular output (variable columns, time at end):
#   0  0.423456  0.312567  0.876543  00:12
_DATA_ROW = re.compile(
    r"^\s*(\d+)\s+"                           # epoch
    r"([\d.eE+-]+)\s+"                        # train_loss
    r"([\d.eE+-]+)\s+"                        # valid_loss
    r"((?:[\d.eE+-]+\s+)*)?"                  # optional extra metrics
    r"(\d{2}:\d{2})\s*$"                      # time
)
# Header detection
_HEADER_LINE = re.compile(r"\btrain_loss\b.*\bvalid_loss\b")


@register_parser
class FastAIParser:
    name = "fastai"
    priority = 75

    def __init__(self) -> None:
        self._extra_headers: list[str] = []

    def sniff(self, sample_lines: list[str]) -> float:
        has_header = any(_HEADER_LINE.search(line) for line in sample_lines)
        has_data = any(_DATA_ROW.match(line) for line in sample_lines)
        if has_header and has_data:
            return 0.90
        if has_header or has_data:
            return 0.40
        return 0.0

    def parse_line(self, line: str, ctx: ParserContext) -> list[RawMetric]:
        if _HEADER_LINE.search(line):
            # Capture extra metric names beyond train_loss / valid_loss
            parts = line.split()
            self._extra_headers = parts[2:-1] if len(parts) > 3 else []
            return []

        m = _DATA_ROW.match(line)
        if not m:
            return []

        ctx.current_epoch = float(m.group(1))
        train_loss = float(m.group(2))
        valid_loss = float(m.group(3))
        extras_raw = (m.group(4) or "").split()

        metrics: list[RawMetric] = [
            RawMetric(
                seq=ctx.seq,
                epoch=ctx.current_epoch,
                key="train_loss",
                value=train_loss,
                parser_name=self.name,
                confidence=0.88,
            ),
            RawMetric(
                seq=ctx.seq,
                epoch=ctx.current_epoch,
                key="valid_loss",
                value=valid_loss,
                parser_name=self.name,
                confidence=0.88,
            ),
        ]

        for i, val_str in enumerate(extras_raw):
            key = self._extra_headers[i] if i < len(self._extra_headers) else f"metric_{i}"
            with contextlib.suppress(ValueError):
                metrics.append(
                    RawMetric(
                        seq=ctx.seq,
                        epoch=ctx.current_epoch,
                        key=key,
                        value=float(val_str),
                        parser_name=self.name,
                        confidence=0.80,
                    )
                )

        return metrics

File: epochix-vscode/src/parsers/fastai.ts
/**
 * TypeScript port of src/epochix/parsers/fastai.py.
 *
 * fastai prints a table, one row per epoch, time last:
 *
 *     epoch  train_loss  valid_loss  accuracy  time
 *     0      0.681724    0.612903    0.671875  00:00
 *
 * The extension had no fastai parser, so the universal one read the time
 * column "00:00" as a metric named `00` and nothing else.
 */
import type { Parser, ParserContext, RawMetric } from "./base";

const DATA_ROW =
  /^\s*(\d{1,9})\s+([\d.eE+-]{1,32})\s+([\d.eE+-]{1,32})\s+((?:[\d.eE+-]{1,32}\s+){0,64})?(\d{2}:\d{2})\s*$/;
const HEADER_LINE = /\btrain_loss\b.*\bvalid_loss\b/;

export class FastAIParser implements Parser {
  readonly name = "fastai";
  readonly priority = 75;
  private _extraHeaders: string[] = [];

  sniff(sampleLines: readonly string[]): number {
    const hasHeader = sampleLines.some((l) => HEADER_LINE.test(l));
    const hasData = sampleLines.some((l) => DATA_ROW.test(l));
    if (hasHeader && hasData) return 0.9;
    if (hasHeader || hasData) return 0.4;
    return 0.0;
  }

  parseLine(line: string, ctx: ParserContext): RawMetric[] {
    if (HEADER_LINE.test(line)) {
      // `epoch train_loss valid_loss <extras…> time`: extras start at index 3.
      const parts = line.trim().split(/\s+/);
      this._extraHeaders = parts.length > 4 ? parts.slice(3, -1) : [];
      return [];
    }
    const m = DATA_ROW.exec(line);
    if (m === null) return [];
    ctx.currentEpoch = parseFloat(m[1]);
    const metric = (key: string, value: number, confidence: number): RawMetric => ({
      seq: ctx.seq,
      epoch: ctx.currentEpoch,
      step: null,
      key,
      value,
      parserName: this.name,
      confidence,
    });
    const out: RawMetric[] = [
      metric("train_loss", parseFloat(m[2]), 0.88),
      metric("valid_loss", parseFloat(m[3]), 0.88),
    ];
    const extras = (m[4] ?? "").trim().split(/\s+/).filter((x) => x.length > 0);
    extras.forEach((raw, i) => {
      const value = parseFloat(raw);
      if (Number.isNaN(value)) return;
      // Validation metrics: fastai computes every column after valid_loss in
      // its validate step (as fastai.py).
      out.push(metric(`valid_${this._extraHeaders[i] ?? `metric_${i}`}`, value, 0.8));
    });
    return out.filter((x) => !Number.isNaN(x.value));
  }
}
